File: src/analyses/embedUMAP.r
library(arrow)
library(data.table)
library(dplyr)
library(uwot)
library(ggplot2)

# Load data
pca <- read_parquet("results/embeddings_full_pca.parquet")
data <- as.data.table(read_parquet("data/processed_data.parquet"))

# Extract PC columns
colnames(pca)
pc_cols <- grep("^PC[0-9]+$", names(pca), value = TRUE)
print(pc_cols)
stopifnot(length(pc_cols) > 0)

# Run UMAP on PC coordinates
set.seed(42)
umap_res <- umap(as.matrix(pca[, pc_cols]), n_neighbors = 15, min_dist = 0.1, n_threads = 4)

# Add UMAP coords to data
pca$UMAP1 <- umap_res[, 1]
pca$UMAP2 <- umap_res[, 2]

# Merge metadata
pca <- pca %>% left_join(data[, .(ID, Organism, Order)], by = "ID")

# Plot by Gene
p_gene <- ggplot(pca, aes(UMAP1, UMAP2, color = Gene)) +
  geom_point(size = 0.1, alpha = 0.3) +
  theme_minimal() +
  ggtitle("UMAP colored by Gene")

# Plot by Residue (site class)
p_residue <- ggplot(pca, aes(UMAP1, UMAP2, color = Residue)) +
  geom_point(size = 0.1, alpha = 0.3) +
  theme_minimal() +
  ggtitle("UMAP colored by Residue")

# Save
ggsave("umap_by_gene.png", p_gene, width = 10, height = 8)
ggsave("umap_by_residue.png", p_residue, width = 10, height = 8)

# If you have significance column, add:
# p_sig <- ggplot(pca, aes(UMAP1, UMAP2, color = significance)) + ...

File: src/analyses/embedUMAP.py
import pandas as pd
import numpy as np
from umap import UMAP
import matplotlib.pyplot as plt

# Load data
pca = pd.read_parquet("results/embeddings_full_pca.parquet")
data = pd.read_parquet("data/processed_data.parquet")

# Extract PC columns
pc_cols = [c for c in pca.columns if c.startswith("PC") and c[2:].isdigit()]
assert len(pc_cols) > 0, "No PC columns found"

# Run UMAP
umap_model = UMAP(n_neighbors=15, min_dist=0.1, random_state=42, n_jobs=-1)
umap_coords = umap_model.fit_transform(pca[pc_cols].values)

pca["UMAP1"] = umap_coords[:, 0]
pca["UMAP2"] = umap_coords[:, 1]

# Merge metadata
pca = pca.merge(data[["ID", "Organism", "Order", "pheno_wc2.1_2.5m_bio_8_p50"]], on="ID", how="left")

# Plot by Gene
fig, ax = plt.subplots(figsize=(10, 8))
for gene in pca["Gene"].unique():
    subset = pca[pca["Gene"] == gene]
    ax.scatter(subset["UMAP1"], subset["UMAP2"], s=0.5, alpha=0.3, label=gene)
ax.set_title("UMAP colored by Gene")
ax.legend(markerscale=5)
plt.savefig("plots/umap_by_gene.png", dpi=150)

# Plot by Residue
fig, ax = plt.subplots(figsize=(10, 8))
for res in pca["Residue"].unique():
    subset = pca[pca["Residue"] == res]
    ax.scatter(subset["UMAP1"], subset["UMAP2"], s=0.5, alpha=0.3, label=res)
ax.set_title("UMAP colored by Residue")
plt.savefig("plots/umap_by_residue.png", dpi=150)

# Plot by phenotype (temperature)
fig, ax = plt.subplots(figsize=(10, 8))
sc = ax.scatter(
    pca["UMAP1"], pca["UMAP2"],
    c=pca["pheno_wc2.1_2.5m_bio_8_p50"],
    s=0.5, alpha=0.3, cmap="viridis"
)
plt.colorbar(sc, label="Temp (BIO8 p50)")
ax.set_title("UMAP colored by Temperature")
plt.savefig("plots/umap_by_temperature.png", dpi=150)